### DISCUSSION TITLE
Suggestions for reliable DHT22/AM2302 readings?

### DISCUSSION THREAD

**Discussion by Xorlent** [HomeSpan/HomeSpan#275] [General]
I've made several projects using the HomeSpan library and all work great.  I've found that with the single core ESP32C3 as soon as I associate the device to Wi-Fi, my DHT22 readings become problematic.  I assume the read timings get skewed due to the WiFi stack using CPU cycles during that process.  I'm wondering if anyone has suggestions for me.  I'm using the DHTNew library and have tried compiling it with interrupts enabled and without -- neither seem to give me truly reliable results.

Thanks in advance!

**Comment by HomeSpan** [comment]
I checked on the Adafruit site and unfortunately the [recommended solution](https://learn.adafruit.com/modern-replacements-for-dht11-dht22-sensors) is not to use the DHT22 since it utilizes such a sensitive and bespoke method for communication.  Instead, they recommend using a [DHT20](https://www.adafruit.com/product/5183) or [AM2301B](https://www.adafruit.com/product/5181) as a replacement since these both communicate using the standard I2C bus protocol (which is natively supported on ESP32 chips).

If you do want to use the DHT22, or any device that requires reading the precise timing of a pulse train, I would recommend using the ESP32's RMT peripheral.  HomeSpan has two classes (RFControl and Pixel) that make use of the RMT peripheral for *transmitting* precisely-timed pulse trains (for things like RF and IR remote controls), but I have not implemented anything anything related to the RMT *receiving* input.  It's definitely possible, but you would basically need to write your own driver using the IDF functions that control the RMT peripheral, which is probably overkill for the DHT22.

**Reply by Xorlent** [reply]
Thanks so much for the guidance, Gregg.  If I do end up cobbling something together, I'll be sure to post the examples.

**Reply by Xorlent** [reply]
Okay, in my testing the following configurations work reliably on an ESP32C3:

DHT22/AM2302 using the DHT_NEW library.  Always check the read result though, because I've found ~ 10% of reads will give an error, but they do work.  For one sensor, poll no more frequently than every ~20 seconds for best results.  For two sensors, poll no more frequently than every ~30 seconds for best results.

If you are also controlling an I2C device via the hardware I2C interface, you cannot use DHT22/AM2302 sensors.

**Reply by TomInIowa** [reply]
I have been using the DS18B20 sensors... onewire communication.

**Reply by Xorlent** [reply]
> I have been using the DS18B20 sensors... onewire communication.

What library did you use and do you also have an I2C device in the same solution?

**Reply by TomInIowa** [reply]
Dallas temperature… it installed other things it needed

**Reply by Xorlent** [reply]
Thanks Tom, it looks like it leverages OneWire, I'll have to try that in conjunction with I2C to see if the issue persists.

**Comment by TomInIowa** [comment]
It see you are working on homespan thermostat.   Did you start from someone's online.  project?  or from scratch.?  I found one but part of it doesn't seem to  work.   No cooling and doesn't display the temperature at the thermostat.

**Comment by Xorlent** [comment]
I built it from scratch and used ss_oled (bit banging) for the display.  I have all features working but as we speak I'm working on revamping the temperature sense hardware setup, since an i2c BMP280 plus DHT22 did not work well together.  I'm thinking of switching everything to bit bang which should cure the issue.  Not sure if this will take me down the wrong path, but I should know soon enough.

**Reply by Xorlent** [reply]
After experimenting some more and trying bit-banging the BMP280, I ended up abandoning the DHT22/AM2302 one wire sensor in this project due to the simultaneous need for I2C.  I found a combo board with AHT20 and BMP280 so all sensors are I2C and it's working great.
